py: Implement very simple frozen modules support.

pfalcon opened the pull request:
Only modules (not packages) supported now. Source modules can be converted
to frozen module structures using tools/make-frozen.py script.

pfalcon commented:
Well, noone wants to implement it, so let's do it in small steps...

dpgeorge commented on py/builtinimport.c:
```diff
@@ -133,6 +132,12 @@ STATIC void do_load(mp_obj_t module_obj, vstr_t *file) {
+STATIC void do_load(mp_obj_t module_obj, vstr_t *file) {
+    // create the lexer
+    mp_lexer_t *lex = mp_lexer_new_from_file(vstr_str(file));
+    do_load_from_lexer(module_obj, lex, vstr_str(file));
```
Is this doing a proper tail call?  Important to keep stack usage down for imports, since they already use a lot of stack.

pfalcon replied:
Yes, it's pretty tail-recursive, and I confirmed that when compiled with optimization on x64, the function ends with jump to do_load_from_lexer.

dpgeorge replied:
I verified it's a jump on stmhal.  And with frozen modules not enabled, it just expands do_load_from_lexer in place here.

dpgeorge commented on py/frozenmod.c:
```diff
@@ -0,0 +1,53 @@
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2014 Damien P. George
```
2015 P Sokolovsky?

dpgeorge commented on py/frozenmod.c:
```diff
@@ -0,0 +1,53 @@
+extern const uint16_t frozen_sizes[];
+extern const char frozen_content[];
+
+mp_lexer_t *find_frozen_module(char *str, int len) {
```
mp_frozen_find_module?

dpgeorge replied:
const char *str?

pfalcon replied:
> mp_frozen_find_module?

Well, I love namespace prefixes, but that sounds too artificial. Let's just have verb at the start for this single function, like many existing functions anyway?

dpgeorge commented on py/frozenmod.c:
```diff
@@ -0,0 +1,53 @@
+#if MICROPY_MODULE_FROZEN
+
+extern const uint16_t frozen_sizes[];
+extern const char frozen_content[];
```
mp_frozen_sizes and mp_frozen_content?

pfalcon replied:
Done

dpgeorge commented on py/frozenmod.c:
```diff
@@ -0,0 +1,53 @@
+        int l = strlen(s);
+        if (l == len && !memcmp(str, s, l)) {
+            s += l + 1;
+            mp_lexer_t *lex = mp_lexer_new_from_str_len(MP_QSTR_, s, *sz_ptr, 0);
```
MP_QSTR_ is passed here as the source name.  Is this for a reason?  It will lead to hard-to-find errors, without the filename given.

pfalcon replied:
Lexer errors are generally not expected to happen for frozen modules. I could pass in qstr from builtin_import, but that potentially would make this oversimplified implementation more tangled (how for example this implementation will be extended to support packages? Dunno.).

dpgeorge commented on tools/make-frozen.py:
```diff
@@ -0,0 +1,50 @@
+# frozen/foo.py
+# frozen/bar.py
+#
+# Run script, passing path to the directory abobe:
```
above

pfalcon replied:
Fixed

dpgeorge commented on tools/make-frozen.py:
```diff
@@ -0,0 +1,50 @@
+#
+# Run script, passing path to the directory abobe:
+#
+# ./make-frozen/py frozen > frozen.c
```
make-frozen.py

Commit:
py: Implement very simple frozen modules support.

Only modules (not packages) supported now. Source modules can be converted
to frozen module structures using tools/make-frozen.py script.

dpgeorge commented:
Ok.  Feel free to merge.

dpgeorge commented:
Rebased and merged by @pfalcon in 640e0b221e972f9a2da2d870bf3fc5a93c18f0ec.

stinos commented on py/builtinimport.c:
```diff
@@ -237,6 +242,15 @@ mp_obj_t mp_builtin___import__(mp_uint_t n_args, const mp_obj_t *args) {
+    #if MICROPY_MODULE_FROZEN
+    mp_lexer_t *lex = mp_find_frozen_module(mod_str, mod_len);
+    if (lex != NULL) {
+        module_obj = mp_obj_new_module(MP_OBJ_QSTR_VALUE(module_name));
```
Is it always safe to use MP_OBJ_QSTR_VALUE here instead of mp_obj_str_get_qstr? Just asking because I got bitten by it myself, see #1077

dpgeorge replied:
Probably not.  On line 227 the qstr is extracted from module_name using mp_obj_str_get_qstr, so should probably cache that return value and use it here.

pfalcon replied:
Wel, my general assumption that module name is identifier and thus always qstr, but thinking that it can be called as `__import__`, yes, need to do that.

pfalcon replied:
Fixed in 8064892c9b894f609b270f85a15ef10f4443d714
